### PR TITLE
DBZ-326, DBZ-327

### PR DESCRIPTION
**Pull request by xenji**
The two tickets unintentionally resulted in a higher modification as expected.
@gunnarmorling Sorry, there were a few conflicting changes (could be sorted to the one or the other ticket). I decided not to put effort in keeping them separated, but in producing working software.

- Move cache initialisation to the class members. The cache is per instance of the class and SMTs are bound to the connector in which they are configured. Clearing the cache on stop() seems therefore unnecessary, as we want to reuse the cache for the same connector instance on (re-)start.
- Remove the identity equality checks in favour of Objects.equals(), which are left-hand null-safe. Null-checks are still needed for semantical purposes.
- Logger moved to static member.
- Add new "isNotNull" validation function to Field, in order to allow empty strings as values.
- Topic Replacement and Key Field Replacement now allow empty, but non-null values.
- Fields where commonly used directly with an implicit .toString() call (e.g. in config.getString()). These calls are now replaced with an explicit .name() call on the field to no rely on the implicit logic.
- Moved for-loops to lambda syntax for better readability
- Reorder members and methods. "Static to local" and within that "public to private"

**What is missing:**
- More tests that use a `SourceRecord` (How do I create one that is close or equal to one that DBZ produces?)
- Docs, docs, docs

Signed-off-by: Mario Mueller <mario@xenji.com>

### CODE REVIEW

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java [hunk 1]
```diff
@@ -107,78 +86,154 @@
...
+        );
+
         if (!config.validateAndRecord(configFields, logger::error)) {
             throw new ConnectException("Unable to validate config.");
         }
 
-        topicRegex = Pattern.compile(config.getString(TOPIC_REGEX));
-        topicReplacement = config.getString(TOPIC_REPLACEMENT);
+        topicRegex = Pattern.compile(config.getString(TOPIC_REGEX.name()));
+        topicReplacement = config.getString(TOPIC_REPLACEMENT.name());
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: The use of `name()` is not necessary and in fact is less readable. Note that `Configuration` has methods that take a `Field` object (such as [getString(Field)](https://github.com/debezium/debezium/blob/master/debezium-core/src/main/java/io/debezium/config/Configuration.java#L1018)) that call `name()`.

Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: The use of `name()` is not necessary and in fact is less readable. Note that `Configuration` has methods that take a `Field` object (such as [getString(Field)](https://github.com/debezium/debezium/blob/master/debezium-core/src/main/java/io/debezium/config/Configuration.java#L1018)) that call `name()`.

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java [hunk 2]
```diff
@@ -107,78 +86,154 @@
...
+    private static final Field KEY_FIELD_REPLACEMENT = Field.create("key.field.replacement")
+            .withDisplayName("Key field replacement")
+            .withType(ConfigDef.Type.STRING)
+            .withWidth(ConfigDef.Width.LONG)
+            .withImportance(ConfigDef.Importance.LOW)
+            .withValidation(ByLogicalTableRouter::validateKeyFieldReplacement)
+            .withDescription("The replacement string used in conjunction with " + KEY_FIELD_REGEX.name() +
+                ". This will be used to create the physical table identifier in the record's key.");
+
+    private static final Logger logger = LoggerFactory.getLogger(ByLogicalTableRouter.class);
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: It's a pity that all the "clean-up" is not in a separate commit. The changes themselves are good, but having all in one commit make it much harder to review the actual change.

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java [hunk 3]
```diff
@@ -272,9 +327,7 @@ private Struct updateEnvelope(Schema newEnvelopeSchema, Struct oldEnvelope) {
 
     private Struct updateValue(Schema newValueSchema, Struct oldValue) {
         final Struct newValue = new Struct(newValueSchema);
-        for (org.apache.kafka.connect.data.Field field : oldValue.schema().fields()) {
-            newValue.put(field.name(), oldValue.get(field));
-        }
+        oldValue.schema().fields().forEach(field -> newValue.put(field.name(), oldValue.get(field)));
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: Surely a matter of taste, but don't think it's more readable than before. An actual downside though is that stepping through this loop in a debugger gets more difficult. Anyways, no need to revert, but I'd perfer to not change such things for the sake of it.

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java [hunk 4]
```diff
@@ -46,38 +46,27 @@
...
             .withValidation(Field::isRequired, Field::isRegex)
             .withDescription("The regex used for extracting the name of the logical table from the original topic name.");
-
     private static final Field TOPIC_REPLACEMENT = Field.create("topic.replacement")
             .withDisplayName("Topic replacement")
             .withType(ConfigDef.Type.STRING)
             .withWidth(ConfigDef.Width.LONG)
             .withImportance(ConfigDef.Importance.LOW)
-            .withValidation(Field::isRequired)
+            .withValidation(Field::isNotNull)
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: How comes there's no counterpart to the logic from `validateKeyFieldReplacement()`? I'd have expected the same here.

Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: Why this change? It seems that the only difference between `isRequired` and `isNotNull` is that the latter allows a zero-length string. Is that intentional to allow a zero-length string? If so, that needs to be documented as to what that means.

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java [hunk 5]
```diff
@@ -107,78 +86,154 @@
...
         final String newTopic = determineNewTopic(oldTopic);
+
         if (newTopic == null) {
             return record;
         }
 
-        logger.debug("Applying topic name transformation from " + oldTopic + " to " + newTopic + ".");
+        if (logger.isDebugEnabled()) {
+            logger.debug("Applying topic name transformation from {} to {}", oldTopic, newTopic);
+        }
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: Good change to parameterize the log statement rather than build up a string. However, the only time it is advantageous to wrap a log statement in an if-enabled check is if computing the parameters to the log statement are not trivial. But they are trivial in this case, so no if-enabled check is required.

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java [hunk 6]
```diff
@@ -242,7 +297,7 @@ private Schema updateEnvelopeSchema(Schema oldEnvelopeSchema, String newTopicNam
         for (org.apache.kafka.connect.data.Field field : oldEnvelopeSchema.fields()) {
             final String fieldName = field.name();
             Schema fieldSchema = field.schema();
-            if (fieldName == Envelope.FieldName.BEFORE || fieldName == Envelope.FieldName.AFTER) {
+            if (Objects.equals(fieldName, Envelope.FieldName.BEFORE) || Objects.equals(fieldName, Envelope.FieldName.AFTER)) {
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: I think object equality actually would have been fine when using this with any of the Debezium connectors, since the Debezium connectors build the `Field` object using the `Envelope`'s actual field name string objects. But I agree this is safer if not more correct.

File: debezium-core/src/test/java/io/debezium/transforms/ByLogicalTableRouterTest.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    public void testBrokenTopicReplacementConfiguration() {
+        final ByLogicalTableRouter<SourceRecord> subject = new ByLogicalTableRouter<>();
+        final Map<String, String> props = new HashMap<>();
+
+        props.put("topic.regex", "someValidRegex(.*)");
+        subject.configure(props);
+    }
+
+    // FIXME: This SMT can use more tests for more detailed coverage.
+    // The creation of a SourceRecord is required for each test
```
Comment (rhauch) [inline debezium-core/src/test/java/io/debezium/transforms/ByLogicalTableRouterTest.java]: Creating a `SourceRecord` is not a big deal ... they're regular POJOs. See some of the [unit tests](https://github.com/apache/kafka/blob/trunk/connect/transforms/src/test/java/org/apache/kafka/connect/transforms/InsertFieldTest.java#L102) in the Kafka project for examples.

File: debezium-core/src/test/java/io/debezium/transforms/ByLogicalTableRouterTest.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+public class ByLogicalTableRouterTest {
+
+    @Test(expected = ConnectException.class)
+    public void testBrokenKeyReplacementConfiguration() {
+        final ByLogicalTableRouter<SourceRecord> subject = new ByLogicalTableRouter<>();
+        final Map<String, String> props = new HashMap<>();
+
+        props.put("topic.regex", "someValidRegex(.*)");
+        props.put("topic.replacement", "$1");
+        props.put("key.field.regex", "If this is set, key.field.replacement must be non-null");
```
Comment (rhauch) [inline debezium-core/src/test/java/io/debezium/transforms/ByLogicalTableRouterTest.java]: Just "non-null" or also "non-empty"?

File: debezium-core/src/main/java/io/debezium/config/Field.java [hunk 9]
```diff
@@ -1175,4 +1175,13 @@ public static int isDouble(Configuration config, Field field, ValidationOutput p
         return 0;
     }
 
-}
\ No newline at end of file
+    public static int isNotNull(Configuration config, Field field, Field.ValidationOutput problems) {
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/config/Field.java]: It'd be nice to have JavaDoc for this method. Granted, the others don't yet so understand that you're just following the existing code. But in this case, it'd be useful to know how this method differs from `isRequired`.

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java (hunk 3 above)
Comment (xenji) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: No problem to revert it. I obey any preferences for the sake of clarity.

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java (hunk 4 above)
Comment (xenji) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: You're right. There is a dependency chain from `topic.regex -> topic.replacement` and, if set, from `key.regex -> key.replacement`. `key.regex` cannot live without `topic.regex` either (makes no sense).

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java (hunk 1 above)
Comment (xenji) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: I must have missed that signature, sorry. I fix that.

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java (hunk 4 above)
Comment (xenji) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: Just recognised my own confusion here. Thanks for catching this. It is required, as the support for the use of back-references in the regex removes the need of an empty string. Just match what you want and make the replacement `$1`.

File: debezium-core/src/test/java/io/debezium/transforms/ByLogicalTableRouterTest.java (hunk 7 above)
Comment (xenji) [inline debezium-core/src/test/java/io/debezium/transforms/ByLogicalTableRouterTest.java]: Right, but there is yet no convenient way to create a `SourceRecord` in the same manner as DBZ emits it, right? The transformation is very much based on the structure of DBZ and therefore a normal `SourceRecord` won't do the trick in the test.

File: debezium-core/src/test/java/io/debezium/transforms/ByLogicalTableRouterTest.java (hunk 8 above)
Comment (xenji) [inline debezium-core/src/test/java/io/debezium/transforms/ByLogicalTableRouterTest.java]: Updated the test to reflect the "truth".

File: debezium-core/src/test/java/io/debezium/transforms/ByLogicalTableRouterTest.java (hunk 7 above)
Comment (rhauch) [inline debezium-core/src/test/java/io/debezium/transforms/ByLogicalTableRouterTest.java]: See the `Envelope` methods that will construct a create, update, or delete envelope around  Struct objects for before, after, and source. You could just create Struct objects and Schemas that work for your tests. But, no, we don't have test utilities in core for constructing SourceRecords.

File: debezium-core/src/main/java/io/debezium/config/Field.java [hunk 10]
```diff
@@ -1175,4 +1175,19 @@ public static int isDouble(Configuration config, Field field, ValidationOutput p
...
 
-}
\ No newline at end of file
+    /**
+     * Validates the field only for being a not-null value.
+     *
+     * The validation here is more relaxed than in {@link Field::isRequired}. In contrast to {@link Field::isRequired},
+     * this method does not check if the string has more than zero characters.
+     */
+    public static int isNotNull(Configuration config, Field field, Field.ValidationOutput problems) {
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/config/Field.java]: Thanks for the JavaDoc. This PR no longer needs this method, so should we keep it?

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java [hunk 11]
```diff
@@ -285,10 +339,12 @@ private SchemaBuilder copySchemaExcludingName(Schema source, SchemaBuilder build
     private SchemaBuilder copySchemaExcludingName(Schema source, SchemaBuilder builder, boolean copyFields) {
         builder.version(source.version());
         builder.doc(source.doc());
-        Map<String, String> params = source.parameters();
+
+        final Map<String, String> params = source.parameters();
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: Is this change necessary? Just trying to minimize the number of changes.

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java [hunk 12]
```diff
@@ -24,20 +22,22 @@
...
-import io.debezium.config.Configuration;
-import io.debezium.config.Field;
-import io.debezium.data.Envelope;
-import io.debezium.util.AvroValidator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static org.apache.kafka.connect.transforms.util.Requirements.requireStruct;
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: I don't think it's a good idea to reorder the imports. They used to match the order used in other classes (e.g., statics, JDK, other libs, Debezium's), so with these changes the next person to modify the class will likely reorder again if their IDE preferences are different. Generally it's better to try to match the preferences already used in the project.

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java [hunk 13]
```diff
@@ -24,20 +22,22 @@
...
  * A logical table consists of one or more physical tables with the same schema. A common use case is sharding -- the
  * two physical tables `db_shard1.my_table` and `db_shard2.my_table` together form one logical table.
- *
+ * <p>
  * This Transformation allows us to change a record's topic name and send change events from multiple physical tables to
  * one topic. For instance, we might choose to send the two tables from the above example to the topic
  * `db_shard.my_table`. The config options {@link #TOPIC_REGEX} and {@link #TOPIC_REPLACEMENT} are used
  * to change the record's topic.
- *
+ * <p>
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: We haven't really been very strict about whether or not we used `<p>`, but unless we'd like to standardize I'm not sure it's worth adding these changes here.

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java (hunk 11 above)
Comment (xenji) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: No, it just states the intention and I added it for semantic clarity. Will remove it.

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java (hunk 13 above)
Comment (xenji) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: That was my IDE, the change was unintentional.

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java (hunk 12 above)
Comment (xenji) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: I revert it to the original order.

File: debezium-core/src/main/java/io/debezium/config/Field.java (hunk 10 above)
Comment (xenji) [inline debezium-core/src/main/java/io/debezium/config/Field.java]: I remove it, fewer code to maintain is better ;)

File: debezium-core/src/main/java/io/debezium/config/Field.java [hunk 14]
```diff
@@ -1174,5 +1174,4 @@ public static int isDouble(Configuration config, Field field, ValidationOutput p
         }
         return 0;
     }
-
-}
\ No newline at end of file
+}
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/config/Field.java]: If you happen to make any other changes, might revert this change as well.

Comment (xenji) [inline debezium-core/src/main/java/io/debezium/config/Field.java]: done

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java (hunk 13 above)
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: On this one we should add the "<p>" as otherwise paragraphs won't be properly separated in rendered output, e.g. when hovering over a comment in Eclipse. +1 for handling it separately, though. I've filed https://issues.jboss.org/browse/DBZ-331 for adding this kind of information to CONTRIBUTING.md.

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java [hunk 15]
```diff
@@ -127,58 +164,76 @@ public void configure(Map<String, ?> props) {
...
+                newKey,
+                newEnvelopeSchema,
+                newEnvelope,
+                record.timestamp()
+        );
     }
 
     @Override
     public void close() {
-        keySchemaUpdateCache = null;
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: How comes you removed that?

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java [hunk 16]
```diff
@@ -127,58 +164,76 @@ public void configure(Map<String, ?> props) {
         if (keyFieldRegexString != null) {
             keyFieldRegexString = keyFieldRegexString.trim();
         }
-        if (keyFieldRegexString != null && keyFieldRegexString != "") {
+        if (keyFieldRegexString != null && !Objects.equals(keyFieldRegexString, "")) {
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: I'd use `!keyFieldRegexString.isEmpty()`.

File: debezium-core/src/test/java/io/debezium/transforms/ByLogicalTableRouterTest.java (hunk 8 above)
Comment (gunnarmorling) [inline debezium-core/src/test/java/io/debezium/transforms/ByLogicalTableRouterTest.java]: Hum, it really should say "non-empty" above, right?

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java [hunk 17]
```diff
@@ -107,15 +87,72 @@
...
+    private String topicReplacement;
+    private Pattern keyFieldRegex;
+    private String keyFieldReplacement;
+    private String keyFieldName;
+    private final Cache<Schema, Schema> keySchemaUpdateCache = new SynchronizedCache<>(new LRUCache<Schema, Schema>(16));
+    private final Cache<Schema, Schema> envelopeSchemaUpdateCache = new SynchronizedCache<>(new LRUCache<Schema, Schema>(16));
+
+    /**
+     * Logic as follows:
+     * If KEY_FIELD_REGEX has a value that is really a regex, then the KEY_FIELD_REPLACEMENT must be a non-null value.
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: s/non-null/non-empty/.

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java [hunk 18]
```diff
@@ -107,15 +87,72 @@
...
+        String keyFieldReplacement = config.getString(KEY_FIELD_REPLACEMENT);
+        if (keyFieldReplacement != null) {
+            keyFieldReplacement = keyFieldReplacement.trim();
+        }
+
+        if (keyFieldRegex == null && keyFieldReplacement == null) {
+            return 0;
+        }
+
+        if (!Objects.equals(keyFieldRegex, "")
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: I'd use `isEmpty()`.

Comment (xenji) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: That would mean that `keyFieldRegex` is guaranteed to be not null, which it is not.

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java (hunk 15 above)
Comment (xenji) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: I moved the cache instance creation to the respective member and made it final. There is one distinct SMT instance per connector (a second, third, ... per connector is theoretically possible, but rather useless). This leads to a potentially warm cache after e.g. a connector restart due to a reconfiguration. Otherwise every config change would invalidate the cache, independent of the SMT being affected by the config change or not.

File: debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java (hunk 18 above)
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java]: Ok.

File: debezium-core/src/test/java/io/debezium/transforms/ByLogicalTableRouterTest.java (hunk 8 above)
Comment (xenji) [inline debezium-core/src/test/java/io/debezium/transforms/ByLogicalTableRouterTest.java]: Classic... changed the one line, but not the other :)

**Comment by xenji** [comment]
Rewording in Jira is done, changes also. Pushing the squash commit now.

**Comment by gunnarmorling** [comment]
Nice work! It's applied, added one more commit to simplify the validation implementation a bit.

Thanks a lot, @xenji!
